### ISSUE TITLE
Restoring service states on OpenVPN install?

### ISSUE THREAD

**Issue by mattock** [OpenVPN/openvpn-build#74]
The current [OpenVPN Chocolatey package](https://chocolatey.org/packages/openvpn) contains Powershell code that [checks](https://github.com/wget/chocolatey_package_openvpn/blob/master/tools/chocolateyInstall.ps1#L53) the state of OpenVPN services (stopped, running) and restores it after install. As discussed [here](https://github.com/wget/chocolatey_package_openvpn/issues/2) it would probably make sense for openvpn.nsi to take care of this instead, so that non-chocolatey users would also benefit from this.

Shall I implement this functionality in openvpn.nsi?

**Comment by chipitsine** [comment]
there's a check of running openvpn.exe, which makes a little sense. it seems it should take of "manually started openvpn.exe"

otherwise, we should care of stopping/running services ... only if user checked to reinstall certain service.


all the above makes me sad, it is complicated in either chosen way

**Comment by mattock** [comment]
Looking at openvpn.nsi this is what we do currently:

* _Section -pre:_ stop OpenVPNService and OpenVPNServiceInteractive ([line 204 ->](https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L204))
* _Section SecService:_ uninstall and then install OpenVPNService ([line 208](https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L271))
* _Section -post:_ set OpenVPNServiceInteractive to start on boot, and also launch it ([line 633 ->](https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L633))

Basically the installer resets the state of OpenVPNServiceInteractive on every install (ensure it's enabled and running), and stops OpenVPNService without restarting it. I think this could definitely be improved if we checked the state (Installed? Running?) of the services in -pre section and restoredthose states in -post section. What happens in the middle does not affect the user that much.

**Comment by wget** [comment]
Actually I was making sure both `OpenVPNServiceInteractive` and `OpenVPNService` were reset to the exact state they were before upgrading the chocolatey package. This was requested by an user (and my own use case as well) relying on OpenVPN to establish an RDP connection to his Windows Server. By default as the installer was not resetting the service states, this led to basically cutting the branch on which we were sitting...
+1 for upstream implementation (i.e. in the NSIS installer).

**Comment by chipitsine** [comment]
we probably should remove check https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L213-L218  (it might be related to running openvpnservice ... if user did not check to upgrade service)

as for the rest, it makes sense to save/restore


and (unrelated to this ticket) we should wrap openvpn-gui.exe check to "if user selected openvpn-gui, then check whether gui is running already"

**Comment by mattock** [comment]
@chipitsine : I may be wrong, but I believe the check on [lines 213-218]( https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L213-L218) is needed to prevents attempts to overwrite executables/libraries that are loaded into memory. Attempts to overwrite such files probably fails and leaves the system in an inconsistent state. Also, the openvpn.exe check is harmless _if_ we restore service state afterwards.

Calling a "restore service state" function in -post _and_ before Quit on [line 217](https://github.com/OpenVPN/openvpn-build/blob/master/windows-nsis/openvpn.nsi#L217) seems like a reasonable approach.

**Comment by chipitsine** [comment]
windows locks files (comparing to linux), you cannot overwrite opened files.

**Comment by mattock** [comment]
> windows locks files (comparing to linux), you cannot overwrite opened files.

Ok, then we need to ensure that there are no running openvpn.exe instances or the install will fail in interesting ways. I will start work on the "restore service state" PR.

**Comment by mattock** [comment]
[This NSI code](http://build.openvpn.net/downloads/temp/servicetest.nsi) implements service state caching and restore:

```
; The name of the installer
Name "ServiceTest"

; The file to write
OutFile "servicetest.exe"

; The default installation directory
InstallDir $PROGRAMFILES\ServiceTest

; Request application privileges for Windows Vista
RequestExecutionLevel admin

ShowInstDetails show

;--------------------------------

!addplugindir .

!include "LogicLib.nsh"

; Pages
Page components
Page directory
Page instfiles

Section "GetServiceState"
  
  ; Set output path to the installation directory.
  SetOutPath $INSTDIR
  
  ; Put file there
  File "ServiceTest.nsi"

  Var /GLOBAL iservice_existed
  Var /GLOBAL iservice_starttype
  Var /GLOBAL iservice_was_running
  Var /GLOBAL service_was_running
  
  SimpleSC::ExistsService "OpenVPNServiceInteractive"
  Pop $iservice_existed
  SimpleSC::GetServiceStartType "OpenVPNServiceInteractive"
  Pop $0
  Pop $iservice_starttype
  SimpleSC::GetServiceStatus "OpenVPNServiceInteractive"
  Pop $0
  Pop $iservice_was_running
  
  SimpleSC::GetServiceStatus "OpenVPNService"
  Pop $0
  Pop $service_was_running
  
  Detailprint "OpenVPNServiceInteractive: exists check = $iservice_existed, status check = $iservice_was_running"
  Detailprint "OpenVPNService: status check = $service_was_running"

 SectionEnd

 Section "StopServices"
	nsExec::ExecToLog '"$SYSDIR\net.exe" stop OpenVPNServiceInteractive /yes'
	nsExec::ExecToLog '"$SYSDIR\net.exe" stop OpenVPNService /yes' 
 SectionEnd 
 
Section "RestoreServiceState"
	; Enable and start OpenVPNServiceInteractive if it did not exist before
	${IfNot} $iservice_existed == 0
		nsExec::ExecToLog '"$SYSDIR\sc.exe" config OpenVPNServiceInteractive start= auto'
		nsExec::ExecToLog '"$SYSDIR\net.exe" start OpenVPNServiceInteractive /yes'
	${EndIf}
	
	; Start OpenVPNServiceInteractive if it was running
	${If} $iservice_was_running == 4
		nsExec::ExecToLog '"$SYSDIR\net.exe" start OpenVPNServiceInteractive /yes'
    ${EndIf} 
	
	; Start OpenVPNService if it was running
	${If} $service_was_running == 4
		nsExec::ExecToLog '"$SYSDIR\net.exe" start OpenVPNService /yes'
    ${EndIf} 
SectionEnd
```

Test installer [here](http://build.openvpn.net/downloads/temp/servicetest.exe). Based on brief testing the logic works correctly. The code uses [Simple Service Plugin](http://nsis.sourceforge.net/NSIS_Simple_Service_Plugin) - I did not feel like parsing net.exe output in NSI or doing something equally horrible :smile:

**Comment by chipitsine** [comment]
there are changes in windows 2012, 2016 like "automatic, delayed" start type.
that plugin does not know about it, I'm afraid it is not developing anymore

actually, there are no actively developed service management plugin for nsis

**Comment by selvanair** [comment]
>Test installer here. Based on brief testing the logic works correctly. The code uses Simple Service Plugin - I did not feel like parsing net.exe output in NSI or doing something equally horrible

Then we could use the plugin to start/stop service as well and get rid of the net.exe lines

Is GetServiceStatus output reliable (i.e cannot be 4) if the service is missing? Or do we have to also check the return value for success? 

> ; Enable and start OpenVPNServiceInteractive if it did not exist before
> 	${IfNot} $iservice_existed == 0
> 		nsExec::ExecToLog '"$SYSDIR\sc.exe" config OpenVPNServiceInteractive start= auto'
> 		nsExec::ExecToLog '"$SYSDIR\net.exe" start OpenVPNServiceInteractive /yes'
> 	${EndIf}

The comment and the code says opposite things? Anyway, we want to install and start IService by default, so the start/restart should happen unless it was present but stopped before install, isn't it? So the comment looks right, not the code. 

Also we should reset the start type to $iservice_starttype if it existed. Same for openvpnservice.

**Comment by selvanair** [comment]
> there are changes in windows 2012, 2016 like "automatic, delayed" start type.
> that plugin does not know about it, I'm afraid it is not being developed anymore

The plugin queries service start type using QueryServiceConfig() which works on later versions on WIndows so the start type we get from the plugin should be a valid value on all supported versions of Windows.  It returns values of various preprocessor defines like SERVICE_AUTO_START (= 0x2) which are in windows header files so microsoft will not/cannot change those for compatibility reasons.

Additional details like delayed start (supported by Vista+?) needs to be queried by a separate function QueryServiceConfig2() which may not be implemented in the plugin, but its not a big deal. None of these services should be set to delayed start, anyway.

**Comment by mattock** [comment]
> Is GetServiceStatus output reliable (i.e cannot be 4) if the service is missing? Or do we have to also check the return value for success?

In my tests GetServiceStatus did not return 4 if the service was missing. To keep things simple I chose not check for the PENDING states - at this point at least.

> The comment and the code says opposite things? Anyway, we want to install and start IService by default, so the start/restart should happen unless it was present but stopped before install, isn't it? So the comment looks right, not the code.

No, both comment and the code are correct. The idea is that _if_ OpenVPNService did not exist at all when the install started, then we know the user has not customized its service settings like StartupType, and we can thus safely use our own (sane) defaults.

**Comment by selvanair** [comment]
Aha,  $iservice_existed == 0 means service existed... My brain went nuts with that and the double negative :) And, with the verb "existed" in the variable I just assumed otherwise...

Let's get this into the installer and test that.

**Comment by mattock** [comment]
I agree the code looks confusing. I will make sure the final code comments explain the idea clearly, otherwise it will get reported as a bug eventually :smile:

**Comment by chipitsine** [comment]
I'm going to add intaller testing to openvpn-powershell-test, we can test few things

- install/uninstall/install (it makes sence, because services are marked as "pending remove" during unstall)
- install to custom folder
- install / start services / reinstall (check state is restored)

?

**Comment by mattock** [comment]
@chipitsine : makes sense.

**Comment by mattock** [comment]
Fixed in PR https://github.com/OpenVPN/openvpn-build/issues/80.
